File: mac/Sources/DMShot/FrameGeometry.swift
import CoreGraphics

/// Pure layout math for the pretty-background frame. No drawing, no AppKit.
/// Mirrored in `windows/DMShot/Editor/FrameGeometry.cs`.
enum FrameGeometry {
    /// Symmetric padding in whole pixels (≥1 when the preset is non-zero).
    static func padding(innerSize: CGSize, padding p: FramePadding) -> CGFloat {
        let longer = max(innerSize.width, innerSize.height)
        let raw = (longer * FramePresets.paddingFraction(p)).rounded()
        return max(1, raw)
    }

    static func outerSize(innerSize: CGSize, padding p: FramePadding) -> CGSize {
        let pad = padding(innerSize: innerSize, padding: p)
        return CGSize(width: innerSize.width + 2 * pad, height: innerSize.height + 2 * pad)
    }

    /// The screenshot's rect inside the outer box (origin at the padding offset).
    static func innerRect(innerSize: CGSize, padding p: FramePadding) -> CGRect {
        let pad = padding(innerSize: innerSize, padding: p)
        return CGRect(x: pad, y: pad, width: innerSize.width, height: innerSize.height)
    }

    /// Corner radius in whole pixels (0 when the preset is None).
    static func cornerRadius(innerSize: CGSize, corner c: FrameCorner) -> CGFloat {
        let frac = FramePresets.cornerFraction(c)
        guard frac > 0 else { return 0 }
        let shorter = min(innerSize.width, innerSize.height)
        return max(1, (shorter * frac).rounded())
    }

    /// Expand an image-space inner rect (crop or full image) by the padding —
    /// the live canvas uses this as its content extent so zoom/pan fit the frame.
    static func outerRect(inner: CGRect, padding p: FramePadding) -> CGRect {
        let pad = padding(innerSize: inner.size, padding: p)
        return inner.insetBy(dx: -pad, dy: -pad)
    }

    static func blurRadius(innerSize: CGSize) -> CGFloat {
        let shorter = min(innerSize.width, innerSize.height)
        return max(1, shorter * FramePresets.blurRadiusFraction)
    }
}

File: windows/DMShot/Editor/FrameGeometry.cs
using System;
using System.Windows;

namespace DMShot.Editor;

/// <summary>Pure layout math for the pretty-background frame. Mirrors
/// mac/Sources/DMShot/FrameGeometry.swift.</summary>
public static class FrameGeometry
{
    public static double Padding(Size innerSize, FramePadding p)
    {
        double longer = Math.Max(innerSize.Width, innerSize.Height);
        double raw = Math.Round(longer * FramePresets.PaddingFraction(p), MidpointRounding.AwayFromZero);
        return Math.Max(1, raw);
    }

    public static Size OuterSize(Size innerSize, FramePadding p)
    {
        double pad = Padding(innerSize, p);
        return new Size(innerSize.Width + 2 * pad, innerSize.Height + 2 * pad);
    }

    public static Rect InnerRect(Size innerSize, FramePadding p)
    {
        double pad = Padding(innerSize, p);
        return new Rect(pad, pad, innerSize.Width, innerSize.Height);
    }

    public static double CornerRadius(Size innerSize, FrameCorner c)
    {
        double frac = FramePresets.CornerFraction(c);
        if (frac <= 0) return 0;
        double shorter = Math.Min(innerSize.Width, innerSize.Height);
        return Math.Max(1, Math.Round(shorter * frac, MidpointRounding.AwayFromZero));
    }

    public static Rect OuterRect(Rect inner, FramePadding p)
    {
        double pad = Padding(inner.Size, p);
        return new Rect(inner.X - pad, inner.Y - pad, inner.Width + 2 * pad, inner.Height + 2 * pad);
    }

    public static double BlurRadius(Size innerSize)
    {
        double shorter = Math.Min(innerSize.Width, innerSize.Height);
        return Math.Max(1, shorter * FramePresets.BlurRadiusFraction);
    }
}
